### PR TITLE
Refactor: Extension UI and Service Worker.

### PR DESCRIPTION
**Pull request by amovar18**
## Description
This PR aims to improve the developer experience by breaking the code into small pieces and increasing the readability of the code.

## Relevant Technical Choices
- For cookies landing page components are exported from the `design-system` assembled in the extension and then passed down the landing page as children.
- The service worker now reuses most of the code to parse headers.
- The Service worker uses constant for message passing types.
- Tests have been added for utilities to catch edge cases.

## Screenshot/Screencast

---

## Checklist

- [ ] I have thoroughly tested this code to the best of my abilities.
- [ ] I have reviewed the code myself before requesting a review.
- [ ] This code is covered by unit tests to verify that it works as intended.
- [ ] The QA of this PR is done by a member of the QA team (to be checked by QA).

### CODE REVIEW

File: packages/extension/src/view/popup/stateProviders/settings/settingsProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+    if (Object.keys(currentSettings).includes('isUsingCDP')) {
+      _setIsUsingCDP(currentSettings.isUsingCDP);
+    }
+  }, []);
+
+  const setUsingCDP = useCallback((newValue: boolean) => {
+    setIsUsingCDPForSettingsDisplay(newValue);
+    chrome.storage.session.set({
```
Comment (ayushnirwal) [inline packages/extension/src/view/popup/stateProviders/settings/settingsProvider.tsx:87]: Isn't this a promise? I think this should be "awaited".

File: packages/extension/src/view/popup/stateProviders/settings/settingsProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        Object.keys(changes).includes('isUsingCDP') &&
+        Object.keys(changes.isUsingCDP).includes('newValue')
+      ) {
+        _setIsUsingCDP(changes?.isUsingCDP?.newValue);
+      }
+    },
+    []
+  );
+
+  const setSettingsChanged = useCallback((newValue: boolean) => {
```
Comment (ayushnirwal) [inline packages/extension/src/view/popup/stateProviders/settings/settingsProvider.tsx:142]: Can we just expose _setSettingsChanged directly?

Comment (amovar18) [reply packages/extension/src/view/popup/stateProviders/settings/settingsProvider.tsx]: Resolved in [8cf3cca](https://github.com/GoogleChromeLabs/ps-analysis-tool/pull/519/commits/8cf3cca2bf225352a90a2884d273273426270151)

File: packages/extension/src/view/popup/stateProviders/settings/settingsProvider.tsx (hunk 1 above)
Comment (amovar18) [reply packages/extension/src/view/popup/stateProviders/settings/settingsProvider.tsx]: Resolved in [4ee3a41](https://github.com/GoogleChromeLabs/ps-analysis-tool/pull/519/commits/4ee3a417178866e7763cc9638d18ccd03c5a8f5c)

File: packages/extension/src/view/popup/stateProviders/cookie/cookieProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+      setOnChromeUrl(true);
+    } else {
+      setOnChromeUrl(false);
+    }
+
+    setTabId(tab[0].id);
+  }, [tabToRead]);
+
+  useEffect(() => {
+    chrome.runtime.sendMessage({
```
[the remark this answers is not in the document]
Comment (amovar18) [reply packages/extension/src/view/popup/stateProviders/cookie/cookieProvider.tsx:114]: It returns a void promise. It will only throw an error if the service-worker file is not available.

**Review by ayushnirwal — APPROVED** [review]

**Comment by mohdsayed** [comment]
@mayan-000 This PR refactors a lot of view part, please review.

File: packages/extension/src/view/devtools/components/cookies/cookieLanding/blockedCookiesSection.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  CookiesLandingContainer,
+  CookiesMatrix,
+  prepareCookieStatsComponents,
+  prepareCookiesCount,
+} from '@ps-analysis-tool/design-system';
+/**
+ * Internal dependencies
+ */
+import { useCookie, useSettings } from '../../../stateProviders';
+import type { DataMapping } from '@ps-analysis-tool/design-system/src/components/cookiesLanding/landingHeader';
```
Comment (mohdsayed) [inline packages/extension/src/view/devtools/components/cookies/cookieLanding/blockedCookiesSection.tsx:30]: Missed?

File: packages/extension/src/constants.ts [hunk 5]
```diff
@@ -31,3 +31,16 @@ export const SETTING_PAGE_CONTROLS = [
       "The PSAT tool is designed for efficient single-tab analysis. While <a class='text-bright-navy-blue dark:text-jordy-blue' href='https://github.com/GoogleChromeLabs/ps-analysis-tool/wiki/PSAT-Settings-and-Permissions#multi-tab-debugging' target='_blank'>multi-tab debugging</a> is available for more comprehensive analysis, it is intended for examining 2-3 tabs simultaneously. Using more tabs may impact the tool's responsiveness.",
   },
 ];
+
+export const DEVTOOLS_CLOSE = 'DEVTOOLS_STATE_CLOSE';
+export const DEVTOOLS_OPEN = 'DEVTOOLS_STATE_OPEN';
+export const POPUP_CLOSE = 'POPUP_STATE_CLOSE';
+export const POPUP_OPEN = 'POPUP_STATE_CLOSE';
```
Comment (mayan-000) [inline packages/extension/src/constants.ts:38]: Typo
```suggestion
export const POPUP_OPEN = 'POPUP_STATE_OPEN';
```

File: packages/extension/src/view/popup/stateProviders/settings/settingsProvider.tsx [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+      });
+      setSettingsChanged(false);
+    }
+  }, [settingsChanged]);
+
+  const changeSyncStorageListener = useCallback(
+    (changes: { [key: string]: chrome.storage.StorageChange }) => {
+      if (
+        Object.keys(changes).includes('allowedNumberOfTabs') &&
+        Object.keys(changes.allowedNumberOfTabs).includes('newValue')
```
Comment (mayan-000) [inline packages/extension/src/view/popup/stateProviders/settings/settingsProvider.tsx:127]:
```suggestion
      if (
        changes?.['allowedNumberOfTabs']?.['newValue']
```

Can be used if covers the check, there are other places also in file.

File: packages/extension/src/view/popup/stateProviders/cookie/cookieProvider.tsx [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    }, 6500);
+
+    return () => {
+      if (loadingTimeout.current) {
+        clearTimeout(loadingTimeout.current);
+      }
+    };
+  }, []);
+
+  const setDebouncedStats = useDebouncedCallback((value) => {
```
Comment (mayan-000) [inline packages/extension/src/view/popup/stateProviders/cookie/cookieProvider.tsx:81]: I guess it's not used now, can be removed if true.

File: packages/cli-dashboard/src/components/siteReport/tabs/cookies/cookiesLandingContainer/cookieLanding/cookiesSection.tsx [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            <MessageBox
+              headerText="No cookies found on this page"
+              bodyText="Please try reloading the page"
+            />
+          ))}
+      <CookiesMatrix
+        tabCookies={affectedCookies}
+        componentData={
+          prepareCookieStatsComponents(prepareCookiesCount(affectedCookies))
+            .legend
```
Comment (mayan-000) [inline packages/cli-dashboard/src/components/siteReport/tabs/cookies/cookiesLandingContainer/cookieLanding/cookiesSection.tsx:67]:
```suggestion
          cookieStatsComponent.legend
```

Affected Cookies is not required in this component.

File: packages/cli-dashboard/src/components/siteReport/tabs/cookies/cookiesLandingContainer/index.tsx [hunk 9]
```diff
@@ -51,33 +46,12 @@ const CookiesLandingContainer = ({
...
-              }
-              tabFrames={tabFrames}
-              description=""
-              showInfoIcon={false}
-              showHorizontalMatrix={false}
-              allowExpand={false}
-            />
-          </div>
-        </div>
+      <CookiesLanding>
```
Comment (mayan-000) [inline packages/cli-dashboard/src/components/siteReport/tabs/cookies/cookiesLandingContainer/index.tsx:49]: We have a missing BlockedCookiesSection here.
With blocked reasons and categories
Screenshot 2024-03-05 at 8 23 53 PM

File: packages/extension/src/view/popup/stateProviders/cookie/cookieProvider.tsx (hunk 7 above)
Comment (amovar18) [reply packages/extension/src/view/popup/stateProviders/cookie/cookieProvider.tsx:81]: We still use it in the storage listener.

File: packages/cli-dashboard/src/components/siteReport/tabs/cookies/cookiesLandingContainer/index.tsx (hunk 9 above)
Comment (amovar18) [reply packages/cli-dashboard/src/components/siteReport/tabs/cookies/cookiesLandingContainer/index.tsx:52]: Fix in latest commit.

File: packages/cli-dashboard/src/components/siteReport/tabs/cookies/cookiesLandingContainer/cookieLanding/cookiesSection.tsx (hunk 8 above)
Comment (amovar18) [reply packages/cli-dashboard/src/components/siteReport/tabs/cookies/cookiesLandingContainer/cookieLanding/cookiesSection.tsx]: Fix in latest commits

File: packages/extension/src/view/popup/stateProviders/settings/settingsProvider.tsx (hunk 6 above)
Comment (amovar18) [reply packages/extension/src/view/popup/stateProviders/settings/settingsProvider.tsx]: Fix in latest commits

File: packages/extension/src/view/devtools/components/cookies/cookieLanding/blockedCookiesSection.tsx (hunk 4 above)
Comment (amovar18) [reply packages/extension/src/view/devtools/components/cookies/cookieLanding/blockedCookiesSection.tsx]: Fix in the latest commits

File: packages/extension/src/view/popup/stateProviders/settings/settingsProvider.tsx [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+      if (changes?.['allowedNumberOfTabs']?.['newValue']) {
+        setAllowedNumberOfTabs(changes?.allowedNumberOfTabs?.newValue);
+        setSettingsChanged(true);
+      }
+
+      if (
+        Object.keys(changes).includes('isUsingCDP') &&
+        Object.keys(changes.isUsingCDP).includes('newValue')
+      ) {
+        _setIsUsingCDP(changes?.isUsingCDP?.newValue);
```
[the remark this answers is not in the document]
Comment (amovar18) [reply packages/extension/src/view/popup/stateProviders/settings/settingsProvider.tsx]: Fixed in latest commit.

**Comment by ayushnirwal** [comment]
@amovar18 please resolve conflicts.

**Review by mohdsayed — APPROVED** [review]
